File: diKoding/XamlMvvmBasic/XamlMvvmBasicCs/DelegateCommand.cs
namespace XamlMvvmBasic
{
    using System;
    using System.Windows.Input;

    public class DelegateCommand 
        : ICommand
    {
        private readonly Action actionExecute;

        private readonly Func<Boolean> funcCanExecute;

        public DelegateCommand(
            Action execute)
            : this(execute, () => true)
        {
        }

        public DelegateCommand(
            Action execute, 
            Func<Boolean> canexecute)
        {
            if (execute == null)
            {
                throw new ArgumentNullException(@"execute");
            }

            actionExecute  = execute;
            funcCanExecute = canexecute;
        }

        public event EventHandler CanExecuteChanged;
        
        public void Execute(
            Object p)
        {
            if (this.CanExecute(null))
            {
                this.actionExecute();
            }
        }

        public Boolean CanExecute(
            Object p)
        {
            return this.funcCanExecute == null || this.funcCanExecute();
        }

        public void RaiseCanExecuteChanged()
        {
            if (this.CanExecuteChanged != null)
            {
                this.CanExecuteChanged(this, EventArgs.Empty);
            }
        }
    }
}

File: diKoding/XamlMvvmBasic/XamlMvvmBasicVb/DelegateCommand.vb
Imports System.Windows.Input

Public Class DelegateCommand
    Implements ICommand
    Private ReadOnly _execute As Action

    Private ReadOnly _canExecute As Func(Of [Boolean])

    Public Sub New(execute As Action)
        Me.New(execute, Function() True)
    End Sub

    Public Sub New(execute As Action, canExecute As Func(Of [Boolean]))
        If execute Is Nothing Then
            Throw New ArgumentNullException("execute")
        End If

        _execute = execute
        _canExecute = canExecute
    End Sub

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

    Public Sub Execute(p As [Object]) Implements ICommand.Execute
        If CanExecute(Nothing) Then
            _execute()
        End If
    End Sub

    Public Function CanExecute(p As [Object]) As [Boolean] Implements ICommand.CanExecute
        Return _canExecute Is Nothing OrElse _canExecute()
    End Function

    Public Sub RaiseCanExecuteChanged()
        RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
    End Sub
End Class
